File: src/model/tensorflowRecommendation.js
import tf from '@tensorflow/tfjs'
const axios = require('axios')

// fetch('http://localhost:3000/api/jobpost/filter?limit=20')
//     .then((response) => {
//         response.json()
//     })
//     .then((data) => console.log(data))
//     .catch((error) => console.log(error))

async function getJobposts() {
    try {
        const response = await axios.get('http://localhost:3000/api/jobpost/filter?limit=20')
        let data = await response.data

        data = data.data.map((jobpost) => ({
            jobpostId: jobpost.jobpost_id,
            stacks: jobpost.stacks ? jobpost.stacks.split(',') : null,
        }))

        console.log(data)
        return data
    } catch (error) {
        console.log(error)
    }
}

getJobposts()

let userStacks = [
    { userId: 1, stacks: 'Java' },
    { userId: 1, stacks: 'Javascript' },
    { userId: 1, stacks: 'MySQL' },
    { userId: 2, stacks: 'Unity' },
    { userId: 2, stacks: 'ReactJS' },
    { userId: 3, stacks: 'Docker' },
    { userId: 3, stacks: 'Python' },
    { userId: 3, stacks: 'Javascript' },
]

async function createModel() {
    const model = tf.sequential()

    const tokenizer = new tf.data.text

}


// console.log(jobposts)

// const shuffled = tf.util.shuffle(data)
// const train = shuffled.take(80)
// const test = shuffled.skip(80)

// https://codelabs.developers.google.com/codelabs/tfjs-training-regression/index.html#4
// function createModel() {
//     const model = tf.sequential()

//     model.add(tf.layers.dense({ inputShape: [1], units: 1 }))
//     model.add(tf.layers.embedding())
// }

// function convertToTensor(data) {
//     return tf.tidy(() => {
//         tf.util.shuffle(data)

//         const inputs = data.map(elem => elem.jobpost_id)
//         const stacks = data.map(elem => elem.stacks.split(','))

//         const inputTensor = tf.tensor2d(inputs, [inputs.length, 1])
//         const stackTensor = tf.tensor2d(stacks, [stacks.length, 1])

//         const inputMax = inputTensor.max()
//         const inputMin = inputTensor.min()
//         const stackMax = stackTensor.max()
//         const stackMin = inputTensor.min()

//         const normalizedInputs = inputTensor
//             .sub(inputMin)
//             .div(inputMax.sub(inputMin))
//         const normalizedStacks = inputTensor
//             .sub(stackMin)
//             .div(inputMax.sub(stackMin))

//         return {
//             inputs: normalizedInputs,
//             stacks: normalizedStacks,
//             inputMax,
//             inputMin,
//             stackMax,
//             stackMin,
//         }
//     })
// }

// async function trainModel(model, inputs, labels) {
//     model.compile({
//         optimizer: tf.train.adam(),
//         loss: tf.losses.meanSquaredError,
//         metrics: ['mse'],
//     })

//     const batchSize = 32
//     const epochs = 50

//     return await model.fit(inputs, labels, {
//         batchSize,
//         epochs,
//         shuffle: true,
//         callbacks: tfvis.show.fitCallbacks(
//             { name: 'Training Performance' },
//             ['loss', 'mse'],
//             { height: 200, callbacks: ['onEpochEnd'] }
//         )
//     })
// }

File: src/model/tensorflowRecommendation.py
import os
import pprint
import tempfile

from typing import Dict, Text

import pandas as pd
import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds
import tensorflow_recommenders as tfrs
from tensorflow import keras

import mysql.connector

# Dataset retrieval
conn = mysql.connector.connect(
    host="127.0.0.1",
    user="jin",
    passwd="password",
    database="jobfit"
)

cursor = conn.cursor()
# cursor.execute(
#     '''select j.jobpost_id, stack from jobpost j
#     left join (select jobpost_id, stack from jobpoststack j
#     inner join stack s  on j.stack_id = s.stack_id) j2 on j.jobpost_id = j2.jobpost_id
#     where stack is not null''')
cursor.execute('''select jobpost_id, stack_id from jobpoststack
                where jobpost_id is not null and stack_id is not null''')
result = cursor.fetchall()

# xJobpost = []
# yJobpost = []
# for jobpost in result:
#     if jobpost[1] is not None:
#         stacks = jobpost[1].split(',')
#         for i in range(len(stacks)):
#             xJobpost.append(jobpost[0])
#             yJobpost.append(stacks[i])

jobpostArr = []
jobpostIds = []
for jobpost in result:
    jobpostArr.append([jobpost[0], jobpost[1]])
    jobpostIds.append(jobpost[0])
jobpostArr = np.asarray(jobpostArr).astype(np.float32)
jobpostIds = np.asarray(jobpostIds).astype(np.float32)
dfJobpost = pd.DataFrame(data=jobpostArr, columns=['jobpostId', 'stackId'])
print(dfJobpost)

# cursor.execute('''select u.user_id, stack from user u
#                 left join (select user_id, stack from userstack us
#                 left join stack s on s.stack_id=us.stack_id) t on t.user_id=u.user_id
#                 where stack is not null''')
cursor.execute('''select user_id, stack_id from userstack
                where user_id is not null and stack_id is not null''')
result = cursor.fetchall()

# xUser = []
# yUser = []
# for user in result:
#     if user[1] is not None:
#         stacks = user[1].split(',')
#         for i in range(len(stacks)):
#             xUser.append(user[0])
#             yUser.append(stacks[i])

userArr = []
userIds = []
for user in result:
    userArr.append([user[0], user[1]])
    userIds.append(user[0])
userArr = np.asarray(userArr).astype(np.float32)
userId = np.asarray(userIds).astype(np.float32)
dfUser = pd.DataFrame(data=userArr, columns=['userId', 'stackId'])
print(dfUser)

cursor.close()
conn.close()


jobpostDataset = tf.data.Dataset.from_tensor_slices(jobpostArr)
userDataset = tf.data.Dataset.from_tensor_slices(userArr)

# for x in jobpostDataset.take(1).as_numpy_iterator():
#     pprint.pprint(x)

tf.random.set_seed(42)
shuffled = userDataset.shuffle(
    100_000, seed=42, reshuffle_each_iteration=False)

train = shuffled.take(80_000)
test = shuffled.skip(80_000).take(20_000)


uniqueJobpostIds = np.unique(jobpostIds)
uniqueUserIds = np.unique(userIds)

# print(type(uniqueJobpostIds))

uniqueUserIds = tf.cast(uniqueUserIds, dtype=tf.float32)

# Query tower
embedding_dimension = 32
# userIdsVocab = tf.keras.layers.StringLookup(mask_token=None)
# userIdsVocab.adapt(dfUser['userId'])
# userModel = tf.keras.Sequential([
#     tf.keras.layers.StringLookup(
#         vocabulary=uniqueUserIds, mask_token=None),
#     # We add an additional embedding to account for unknown tokens.
#     tf.keras.layers.Embedding(len(uniqueUserIds) + 1, embedding_dimension)
# ])

# Candidate tower
# jobpostModel = tf.keras.Sequential([
#     tf.keras.layers.StringLookup(
#         vocabulary=uniqueJobpostIds, mask_token=None),
#     tf.keras.layers.Embedding(len(uniqueJobpostIds) + 1, embedding_dimension)
# ])


# class Model(tfrs.Model):
#     def __init__(self,
#                  userModel: tf.keras.Model,
#                  jobpostModel: tf.keras.Model,
#                  task: tfrs.tasks.Retrieval):
#         super().__init__()

#     # Set up user and movie representations.
#     self.userModel = userModel
#     self.jobpostModel = jobpostModel

#     # Set up a retrieval task.
#     self.task = task

#     def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:
#         # Define how the loss is computed.

#         user_embeddings = self.userModel(features["user_id"])
#         movie_embeddings = self.jobpostModel(features["movie_title"])

#         return self.task(user_embeddings, movie_embeddings)


inputShape = (len(dfJobpost.columns,))

# Model definition
model = keras.Sequential([
    keras.layers.Dense(64, activation='relu', input_shape=inputShape),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(len(dfJobpost), activation='sigmoid')
])

# Model compilation
model.compile(loss='binary_crossentropy',
              optimizer='adam', metrics=['accuracy'])
